shadowsocks/udp.go

```diff
@@ -317,13 +351,23 @@ func timedCopy(clientAddr net.Addr, clientConn net.PacketConn, cipher shadowaead
 			if err != nil {
 				return onet.NewConnectionError("ERR_WRITE", "Failed to write to client", err)
 			}
+
+			if atomic.LoadInt32(&targetConn.mode) == modeSingleDNS {
```

Comment:
It will be helpful to have a separation of concerns here in order to prevent complexity from spreading. In particular, it would be cleaner if timedCopy was agnostic about the optimizations. I believe all optimizations can live in natconn.

A few options:
1. return a timeout error in the next ReadFrom call
1. Provide an Expired() method, so the for-loop checks `for !targetConn.Expired()`

The second one may reflect the semantics better.

Reply:
I found a clean implementation of option 1.